Fabric custom ingredient API

Technici4n opened the pull request:
This API allows modders to define custom `Ingredient`s, with more advanced matching logic than what vanilla allows.

Custom ingredients need to implement the `CustomIngredient` interface, and also provide and register a `CustomIngredientDeserializer`. No overriding of `Ingredient` is needed.

Have a look at the API here: https://github.com/Technici4n/fabric/tree/custom-ingredient-api/fabric-recipe-api-v1/src/main/java/net/fabricmc/fabric/api/recipe/v1/ingredient.

The default conditions can be found here: https://github.com/Technici4n/fabric/blob/custom-ingredient-api/fabric-recipe-api-v1/src/main/java/net/fabricmc/fabric/api/recipe/v1/ingredient/DefaultCustomIngredients.java.

Here is how the default conditions are implemented: https://github.com/Technici4n/fabric/tree/custom-ingredient-api/fabric-recipe-api-v1/src/main/java/net/fabricmc/fabric/impl/recipe/ingredient/builtin.

Closes #116.

TODO:
- [x] Decide on exact API and impl structure.
- [x] Add partial/strict NBT, and and difference ingredients.
- [x] JSON and network serialization and deserialization tests.
- [x] Shapeless recipe matching tests.
- [x] `isEmpty` checks tests.
- [x] Write javadoc.
- [x] Test what happens with vanilla clients. See https://github.com/FabricMC/fabric/pull/2586#issuecomment-1355037420
- [x] Rename to Recipe API.

Technici4n committed:
Fabric custom ingredient API

AlphaMode commented:
A possible NBT ingredient would also be useful

Technici4n commented:
> A possible NBT ingredient would also be useful

I have the following ingredients in mind, just haven't written them yet: (matching Forge's)
- strict and partial NBT ingredient
- and ingredient
- difference ingredient

jaredlll08 commented on fabric-ingredient-api-v1/src/main/java/net/fabricmc/fabric/api/ingredient/v1/CustomIngredient.java:
```diff
@@ -0,0 +1,20 @@
+public interface CustomIngredient {
+	boolean matchesStack(ItemStack stack);
+	// TODO: match yarn name? this name derived from old 1.16 yarn Recipe#getPreviewInputs() is more accurate
+	ItemStack[] getPreviewStacks();
```
I feel the name `getPreviewStacks` gives the wrong impression, the method is used in `Recipe#isEmpty` to determine if a recipe is is incomplete.

Technici4n replied:
Renamed to `getMatchingStacks`.

jaredlll08 commented on fabric-ingredient-api-v1/src/main/java/net/fabricmc/fabric/api/ingredient/v1/FabricIngredient.java:
```diff
@@ -0,0 +1,20 @@
+// TODO: ignoresNbt is definitely useful, but what about the other two?
+// TODO: getCustomIngredient might return weird results for "OR" ingredients
+public interface FabricIngredient {
+	default boolean isCustom() {
+		return false;
+	}
+
+	@Nullable
+	default CustomIngredient getCustomIngredient() {
+		return null;
+	}
+
+	default boolean ignoresNbt() {
+		return true;
+	}
```
I'm not too sure how fabric api deals with this, but should these not have the `fabric_` prefix?

Technici4n replied:
For injected interfaces, the status quo is to not add the `fabric_` prefix. Personally. I think that it looks quite ugly, and given fabric API's privileged status we can request other mods to adjust their methods in case of a name conflict.

jaredlll08 commented on fabric-ingredient-api-v1/src/main/java/net/fabricmc/fabric/api/ingredient/v1/CustomIngredient.java:
```diff
@@ -0,0 +1,20 @@
+
+// TODO: what about getMatchingItemIds ?
+public interface CustomIngredient {
+	boolean matchesStack(ItemStack stack);
```
An Ingredient is basically a `Predicate<ItemStack>` so why not have this as
```suggestion
	boolean test(ItemStack stack);
```

jaredlll08 reviewed (commented):
Looking good so far, however there is a pretty big issue that isn't being dealt with by this PR, ShapelessRecipe.

ShapelessRecipe uses `Ingredient#getMatchingItemIds` to determine if a recipe matches or not, and it is pretty fast doing that, however it gains it's speed by not calling `Ingredient#test` and instead uses the integer id of the Item (see `RecipeMatcher#getItemId`), so to make this PR work with ShapelessRecipes, it needs to mixin ShapelessRecipe and change it to call `Ingredient#test`, which is not as simple as that may sound.

A ShapelessRecipe is basically a [Bipartite Graph](https://en.wikipedia.org/wiki/Bipartite_graph) where each input item needs to match an Ingredient, however multiple items may match the same ingredient.

So a recipe like this (In CraftTweaker format since I don't know the json format for nbt):

```zenscript
craftingTable.addShapeless("name", <item:minecraft:dirt>, 
[
<item:minecraft:diamond_sword>.withEnchantment(<enchantment:minecraft:unbreaking>),
<item:minecraft:diamond_sword>.withEnchantment(<enchantment:minecraft:sharpness>)
]);
```

Which requires 2 Diamond Swords, one with Unbreaking and one with Sharpness, needs to make sure that if a player puts:

```zenscript
[
<item:minecraft:diamond_sword>.withEnchantment(<enchantment:minecraft:unbreaking>).withEnchantment(<enchantment:minecraft:sharpness>),
<item:minecraft:diamond_sword>.withEnchantment(<enchantment:minecraft:sharpness>)
]
```

Into a crafting grid, it doesn't try and use the Sword with Unbreaking to fill the Ingredient that asks for Sharpness (assuming partial nbt matching).

To solve this you need to implement something like the [Hopcroft–Karp algorithm](https://en.wikipedia.org/wiki/Hopcroft%E2%80%93Karp_algorithm), which produces how many ItemStacks match the Ingredients.

There may be other solutions to the problem, but I know that Hopcroft-Karp solves it.

Technici4n committed:
More Ingredient API work

Technici4n commented:
Thanks a lot for your detailed explanation.

I decided to go for the simpler augmenting paths algorithm for the maximum cardinality bipartite matching problem, which should look like this : https://github.com/Technici4n/fabric/blob/custom-ingredient-api/fabric-ingredient-api-v1/src/main/java/net/fabricmc/fabric/impl/ingredient/ShapelessMatch.java. I haven't tested it yet though, but I will definitely add tests similar to what you suggested.

I also renamed `matchesStack` and `getPreviewStacks` to `test` and `getMatchingStacks` respectively, and provided some javadoc for `CustomIngredient`. I hope the javadoc is clear enough - there's some tricky details that might not always be obvious. I also borrowed your name for `requiresTesting` which I like better than `ignoresNbt` - hopefully you don't mind. :smile:

Technici4n committed:
Optimize shapeless matching logic

Technici4n committed:
Fix all the things

Technici4n commented:
I just tested the following cases:
- Server with Fabric ingredient API, vanilla client, no recipe using custom ingredients: can connect fine.
- Server with Fabric ingredient API, vanilla client, some recipes using custom ingredients: can't connect, with the following error:
![image](https://user-images.githubusercontent.com/13494793/207747523-ffa718bc-0c38-4998-b5b1-2a912529f249.png)
**Is this fine?**
- Client/server with both Fabric ingredient API and [Faux ingredient extension API](https://github.com/FauxMC/IngredientExtensionAPI): packet writing sending works, but I used the 1.19.2 version of the ingredient extension API, so I get a crash due to other reasons. I think the two APIs are compatible. (Though obviously the ingredient extension API won't be needed anymore once this PR is merged, but I want a smooth transition).

This PR is now ready for reviews!

Technici4n committed:
Move custom ingredient network serialization to account for ingredient extension API

jaredlll08 commented:
> Client/server with both Fabric ingredient API and [Faux ingredient extension API](https://github.com/FauxMC/IngredientExtensionAPI): packet writing sending works, but I used the 1.19.2 version of the ingredient extension API, so I get a crash due to other reasons. I think the two APIs are compatible. (Though obviously the ingredient extension API won't be needed anymore once this PR is merged, but I want a smooth transition

We have no intention of continuing development on the Faux Ingredient Extension API once this PR is merged, so even if it wasn't compatible, it shouldn't be an issue since neither APIs should share a version.

Technici4n commented:
My thinking is that you might release a 1.19.3 version of your Extension API before this PR gets merged in 1.19.3. (I have no idea when it will be merged, it's going to require quite a bit of review).

Patbox commented on fabric-ingredient-api-v1/src/main/java/net/fabricmc/fabric/mixin/ingredient/IngredientMixin.java:
```diff
@@ -0,0 +1,125 @@
+	 * @reason Support custom ingredient network deserialization.
+	 */
+	@Overwrite
+	public static Ingredient fromPacket(PacketByteBuf buf) {
```
This could be replaced with an injection

Technici4n replied:
Could be but I don't see the point since I'm basically overwriting the entire method.

Patbox replied:
You could allow vanilla code to run, for better compat for things that inject there. And for vanilla code reading that, you can set packbufers index to one before reading

Technici4n replied:
So I can "unread" the int? That would work better indeed.

Patbox replied:
Yep, just store old index before reading and restore it if size is different than -1

Technici4n replied:
Done: https://github.com/FabricMC/fabric/pull/2586/commits/9a154dafb484bca11ed4f243b36758756c356ac2.

Patbox commented on fabric-ingredient-api-v1/src/main/java/net/fabricmc/fabric/mixin/ingredient/IngredientMixin.java:
```diff
@@ -0,0 +1,125 @@
+		// Done here instead of writing the ingredient in the subclass for compat with faux ingredient extension API
+		CustomIngredient customIngredient = getCustomIngredient();
+
+		if (customIngredient != null) {
```
You could check with networking api if client has required fabric version for better compatibility. Also it could be used to skip ingredients unsupported by client

Technici4n replied:
I can't skip ingredients that are not supported by the client due to how reloading works, but I can kick clients that don't have the ingredient API if the recipe packet contains custom ingredients. I'll try that.

Patbox replied:
Because that method is written on networking thread, you can get the player, see Nucleoid's Packet Tweaker library

Technici4n replied:
Look at the way reload is implemented. It's using a single packet for all connected clients.

Patbox replied:
Recipes are still written in separate bytebufs per player (packet only contains list of recipes), so that would still work (polymer works thanks to that for example)

Technici4n replied:
Ahhh I see, I'll have a look. Thanks!

Should I just skip invalid recipes, perhaps with a console warning, or disconnect the player?

Patbox replied:
Since fabric contains these by default, just skipping recipes would cause least confusion. alternative would be just writing empty ingredient (0 as size and nothing else). As for warning, it could get rather spammy so maybe only in dev/needs to be enabled by hand?

Patbox replied:
To be fair looking at it now, writing result of getMatchingStacks for client not supporting required type would be also a good alternative

Technici4n replied:
It might lead to ugly desyncs though, where the client thinks it can craft something but the server disagrees. Maybe we should write a barrier block with a warning in the name, Forge-style. :joy:

Patbox replied:
Well, adding some random nbt (with hard to get with commands key/value) would resolve auto matching

Shnupbups commented:
Would it be out of scope to add a way to give these custom ingredients a recipe remainder that overrides the one in their item's settings?
So one could, for example, make a "chocolate milk bucket" recipe that takes a milk bucket and cocoa beans, without the milk bucket leaving behind an empty bucket by overriding the milk bucket's remainder in just that recipe.

Technici4n commented:
> Would it be out of scope to add a way to give these custom ingredients a recipe remainder that overrides the one in their item's settings? So one could, for example, make a "chocolate milk bucket" recipe that takes a milk bucket and cocoa beans, without the milk bucket leaving behind an empty bucket by overriding the milk bucket's remainder in just that recipe.

I think it would be out of scope. Some of the problems:
- Ingredients are used in many places, so properly supporting this would require a lot of patches.
- The patches are very invasive as they require all remainder logic to go through the ingredient.
- For recipes, we don't even know which ingredients were matched where in the getRecipeRemainder method, so we'd need to recompte the match before we apply custom remainders.
I think it's best left to other mods (for crafting recipes, it's "easy" to implement on top of this API with a custom recipe type).

apple502j commented on fabric-ingredient-api-v1/src/main/java/net/fabricmc/fabric/api/ingredient/v1/CustomIngredient.java:
```diff
@@ -0,0 +1,101 @@
+ */
+public interface CustomIngredient {
+	/**
+	 * Check if a stack matches this ingredient.
```
```suggestion
	 * Checks if a stack matches this ingredient.
```

apple502j commented on fabric-ingredient-api-v1/src/main/java/net/fabricmc/fabric/api/ingredient/v1/CustomIngredient.java:
```diff
@@ -0,0 +1,101 @@
+	 * Check if a stack matches this ingredient.
+	 *
+	 * @param stack the stack to test
+	 * @return true if the stack matches this ingredient, false otherwise
```
```suggestion
	 * @return {@code true} if the stack matches this ingredient, {@code false} otherwise
```

apple502j commented on fabric-ingredient-api-v1/src/main/java/net/fabricmc/fabric/api/ingredient/v1/CustomIngredient.java:
```diff
@@ -0,0 +1,101 @@
+	ItemStack[] getMatchingStacks();
+
+	/**
+	 * Return whether this ingredient always requires {@linkplain #test direct stack testing}.
```
Consistency? (Use "Returns" or inline return)

apple502j commented on fabric-ingredient-api-v1/src/main/java/net/fabricmc/fabric/api/ingredient/v1/CustomIngredient.java:
```diff
@@ -0,0 +1,101 @@
+	/**
+	 * Return whether this ingredient always requires {@linkplain #test direct stack testing}.
+	 *
+	 * <p>If false, {@linkplain #test testing this ingredient} with an item stack must be equivalent to checking whether
```
```suggestion
	 * <p>If {@code false}, {@linkplain #test testing this ingredient} with an item stack must be equivalent to checking whether
```

apple502j commented on fabric-ingredient-api-v1/src/main/java/net/fabricmc/fabric/api/ingredient/v1/CustomIngredient.java:
```diff
@@ -0,0 +1,101 @@
+	 * Return whether this ingredient always requires {@linkplain #test direct stack testing}.
+	 *
+	 * <p>If false, {@linkplain #test testing this ingredient} with an item stack must be equivalent to checking whether
+	 * the item stack's item is contained in the ingredient's {@linkplain #getMatchingStacks() list of matching stacks}.
```
```suggestion
	 * the item stack's item is included in the ingredient's {@linkplain #getMatchingStacks() list of matching stacks}.
```

apple502j commented on fabric-ingredient-api-v1/src/main/java/net/fabricmc/fabric/api/ingredient/v1/CustomIngredient.java:
```diff
@@ -0,0 +1,101 @@
+	 * In that case, optimized matching logic can be used, for example using {@link Ingredient#getMatchingItemIds()}.
+	 *
+	 * <p>If true, the ingredient must always be tested using {@link #test(ItemStack)}.
+	 * Note: Fabric patches some vanilla systems such as shapeless recipes to account for this.
```
```suggestion
	 * <p>If {@code true}, the ingredient must always be tested using {@link #test(ItemStack)}.
	 * Note that Fabric patches some vanilla systems such as shapeless recipes to account for this.
```

apple502j commented on fabric-ingredient-api-v1/src/main/java/net/fabricmc/fabric/api/ingredient/v1/CustomIngredient.java:
```diff
@@ -0,0 +1,101 @@
+	 * <p>If true, the ingredient must always be tested using {@link #test(ItemStack)}.
+	 * Note: Fabric patches some vanilla systems such as shapeless recipes to account for this.
+	 *
+	 * @return true if this ingredient ignores NBT data when matching stacks, false otherwise
```
🤦🏼 I think you mean the opposite

apple502j commented on fabric-ingredient-api-v1/src/main/java/net/fabricmc/fabric/api/ingredient/v1/CustomIngredient.java:
```diff
@@ -0,0 +1,101 @@
+	/**
+	 * {@return the serializer for this ingredient}
+	 *
+	 * <p>The serializer must have be registered using {@link CustomIngredientSerializer#register}.
```
```suggestion
	 * <p>The serializer must have been registered using {@link CustomIngredientSerializer#register}.
```

apple502j commented on fabric-ingredient-api-v1/src/main/java/net/fabricmc/fabric/api/ingredient/v1/CustomIngredientSerializer.java:
```diff
@@ -0,0 +1,76 @@
+ */
+public interface CustomIngredientSerializer<T extends CustomIngredient> {
+	/**
+	 * Register a custom ingredient serializer, using the {@linkplain CustomIngredientSerializer#getIdentifier() serializer's identifier}.
```
```suggestion
	 * Registers a custom ingredient serializer, using the {@linkplain CustomIngredientSerializer#getIdentifier() serializer's identifier}.
```

apple502j commented on fabric-ingredient-api-v1/src/main/java/net/fabricmc/fabric/api/ingredient/v1/CustomIngredientSerializer.java:
```diff
@@ -0,0 +1,76 @@
+	}
+
+	/**
+	 * {@return the custom ingredient serializer registered with the given identifier, or {@code null} if none is registered}.
```
```suggestion
	 * {@return the custom ingredient serializer registered with the given identifier, or {@code null} if there is no such serializer}.
```

The current sentence reads like it returns null when there is no serializer at all, which is incorrect.

apple502j commented on fabric-ingredient-api-v1/src/main/java/net/fabricmc/fabric/api/ingredient/v1/CustomIngredientSerializer.java:
```diff
@@ -0,0 +1,76 @@
+	Identifier getIdentifier();
+
+	/**
+	 * Deserialize the custom ingredient from a JSON object.
```
```suggestion
	 * Deserializes the custom ingredient from a JSON object.
```

apple502j commented on fabric-ingredient-api-v1/src/main/java/net/fabricmc/fabric/api/ingredient/v1/CustomIngredientSerializer.java:
```diff
@@ -0,0 +1,76 @@
+	/**
+	 * Deserialize the custom ingredient from a JSON object.
+	 */
+	T read(JsonObject json);
```
Yarn convention is deserialize/serialize. "Read/write" is used for mutable objects reading from/writing to an object.

Technici4n replied:
Then `RecipeSerializer` is wrong.

apple502j replied:
Ah, in that case keep it as is.

apple502j commented on fabric-ingredient-api-v1/src/main/java/net/fabricmc/fabric/api/ingredient/v1/CustomIngredientSerializer.java:
```diff
@@ -0,0 +1,76 @@
+	T read(JsonObject json);
+
+	/**
+	 * Serialize the custom ingredient to a JSON object.
```
```suggestion
	 * Serializes the custom ingredient to a JSON object.
```

apple502j commented on fabric-ingredient-api-v1/src/main/java/net/fabricmc/fabric/api/ingredient/v1/CustomIngredientSerializer.java:
```diff
@@ -0,0 +1,76 @@
+	/**
+	 * Serialize the custom ingredient to a JSON object.
+	 */
+	void write(JsonObject json, T ingredient);
```
See above. Also, should this be `JsonObject write(T ingredient)` instead perhaps?

Technici4n replied:
It's nicer not to need to create a new object imo (and the JsonObject already contains the serializer ID by then, iirc).

apple502j commented on fabric-ingredient-api-v1/src/main/java/net/fabricmc/fabric/api/ingredient/v1/CustomIngredientSerializer.java:
```diff
@@ -0,0 +1,76 @@
+	void write(JsonObject json, T ingredient);
+
+	/**
+	 * Deserialize the custom ingredient from a packet buffer.
```
```suggestion
	 * Deserializes the custom ingredient from a packet buffer.
```

apple502j commented on fabric-ingredient-api-v1/src/main/java/net/fabricmc/fabric/api/ingredient/v1/CustomIngredientSerializer.java:
```diff
@@ -0,0 +1,76 @@
+	T read(PacketByteBuf buf);
+
+	/**
+	 * Serialize the custom ingredient to a packet buffer.
```
```suggestion
	 * Serializes the custom ingredient to a packet buffer.
```

apple502j commented on fabric-ingredient-api-v1/src/main/java/net/fabricmc/fabric/api/ingredient/v1/DefaultCustomIngredients.java:
```diff
@@ -0,0 +1,134 @@
+import net.fabricmc.fabric.impl.ingredient.builtin.OrIngredient;
+
+/**
+ * Factory methods for the custom ingredients directly provided by fabric.
```
```suggestion
 * Factory methods for the custom ingredients directly provided by Fabric API.
```

apple502j commented on fabric-ingredient-api-v1/src/main/java/net/fabricmc/fabric/api/ingredient/v1/DefaultCustomIngredients.java:
```diff
@@ -0,0 +1,134 @@
+ */
+public final class DefaultCustomIngredients {
+	/**
+	 * Create an ingredient that matches what its sub-ingredients all match.
```
```suggestion
	 * Creates an ingredient that matches when its sub-ingredients all match.
```

apple502j commented on fabric-ingredient-api-v1/src/main/java/net/fabricmc/fabric/api/ingredient/v1/DefaultCustomIngredients.java:
```diff
@@ -0,0 +1,134 @@
+	 * }</pre>
+	 */
+	public static Ingredient and(Ingredient... ingredients) {
+		for (Ingredient ing : ingredients) Objects.requireNonNull(ing);
```
Have error message? (Also, should we document that empty is banned?)

apple502j commented on fabric-ingredient-api-v1/src/main/java/net/fabricmc/fabric/api/ingredient/v1/DefaultCustomIngredients.java:
```diff
@@ -0,0 +1,134 @@
+	}
+
+	/**
+	 * Create an ingredient that matches what any of its sub-ingredients matches.
```
```suggestion
	 * Creates an ingredient that matches when any of its sub-ingredients matches.
```

apple502j commented on fabric-ingredient-api-v1/src/main/java/net/fabricmc/fabric/api/ingredient/v1/DefaultCustomIngredients.java:
```diff
@@ -0,0 +1,134 @@
+	 * }</pre>
+	 */
+	public static Ingredient or(Ingredient... ingredients) {
+		for (Ingredient ing : ingredients) Objects.requireNonNull(ing);
```
See above

apple502j commented on fabric-ingredient-api-v1/src/main/java/net/fabricmc/fabric/api/ingredient/v1/DefaultCustomIngredients.java:
```diff
@@ -0,0 +1,134 @@
+	}
+
+	/**
+	 * Create an ingredient that matches if its base ingredient matches, and its subtracted ingredient does NOT match.
```
```suggestion
	 * Creates an ingredient that matches if its base ingredient matches, and its subtracted ingredient <strong>does not</strong> match.
```

Why CAPS when we can **strong**?

apple502j commented on fabric-ingredient-api-v1/src/main/java/net/fabricmc/fabric/api/ingredient/v1/DefaultCustomIngredients.java:
```diff
@@ -0,0 +1,134 @@
+	 */
+	public static Ingredient difference(Ingredient base, Ingredient subtracted) {
+		Objects.requireNonNull(base);
+		Objects.requireNonNull(subtracted);
```
Error message?

apple502j commented on fabric-ingredient-api-v1/src/main/java/net/fabricmc/fabric/api/ingredient/v1/DefaultCustomIngredients.java:
```diff
@@ -0,0 +1,134 @@
+	}
+
+	/**
+	 * Create an ingredient that wraps another ingredient to also check for stack NBT.
```
```suggestion
	 * Creates an ingredient that wraps another ingredient to also check for stack NBT.
```

apple502j commented on fabric-ingredient-api-v1/src/main/java/net/fabricmc/fabric/api/ingredient/v1/DefaultCustomIngredients.java:
```diff
@@ -0,0 +1,134 @@
+	 * Create an ingredient that wraps another ingredient to also check for stack NBT.
+	 * This check can either be strict (the exact NBT must match) or non-strict (the ingredient NBT must be a subset of the stack NBT).
+	 *
+	 * <p>Note: refer to {@link NbtHelper#matches} for the non-strict matching.
```
```suggestion
	 * <p>See {@link NbtHelper#matches} for the non-strict matching.
```

apple502j commented on fabric-ingredient-api-v1/src/main/java/net/fabricmc/fabric/api/ingredient/v1/DefaultCustomIngredients.java:
```diff
@@ -0,0 +1,134 @@
+	 * {
+	 *    "fabric:type": "fabric:nbt",
+	 *    "base": // base ingredient,
+	 *    "nbt": // NBT tag to match, either in JSON directly or a string representation (default: null),
```
Document that null means the absence of NBT in strict, and prohibited in non-strict?

apple502j commented on fabric-ingredient-api-v1/src/main/java/net/fabricmc/fabric/api/ingredient/v1/DefaultCustomIngredients.java:
```diff
@@ -0,0 +1,134 @@
+	 * }</pre>
+	 */
+	public static Ingredient nbt(Ingredient base, @Nullable NbtCompound nbt, boolean strict) {
+		Objects.requireNonNull(base);
```
Error message?

apple502j commented on fabric-ingredient-api-v1/src/main/java/net/fabricmc/fabric/api/ingredient/v1/DefaultCustomIngredients.java:
```diff
@@ -0,0 +1,134 @@
+	}
+
+	/**
+	 * Create an ingredient that matches the passed template stack, NBT included.
```
```suggestion
	 * Creates an ingredient that matches the passed template stack, including NBT.
```

apple502j commented on fabric-ingredient-api-v1/src/main/java/net/fabricmc/fabric/api/ingredient/v1/DefaultCustomIngredients.java:
```diff
@@ -0,0 +1,134 @@
+	 * Create an ingredient that matches the passed template stack, NBT included.
+	 * This check can either be strict (the exact NBT must match) or non-strict (the template NBT must be a subset of the stack NBT).
+	 *
+	 * <p>Note: refer to {@link NbtHelper#matches} for the non-strict matching.
```
```suggestion
	 * <p>See {@link NbtHelper#matches} for the non-strict matching.
```

apple502j commented on fabric-ingredient-api-v1/src/main/java/net/fabricmc/fabric/api/ingredient/v1/DefaultCustomIngredients.java:
```diff
@@ -0,0 +1,134 @@
+	 * <p>Note: refer to {@link NbtHelper#matches} for the non-strict matching.
+	 */
+	public static Ingredient nbt(ItemStack stack, boolean strict) {
+		Objects.requireNonNull(stack);
```
Error message?

apple502j commented on fabric-ingredient-api-v1/src/main/java/net/fabricmc/fabric/api/ingredient/v1/FabricIngredient.java:
```diff
@@ -0,0 +1,42 @@
+	}
+
+	/**
+	 * @see CustomIngredient#requiresTesting()
```
Should this document the behavior for vanilla?

apple502j commented on fabric-ingredient-api-v1/src/main/java/net/fabricmc/fabric/impl/ingredient/CustomIngredientImpl.java:
```diff
@@ -0,0 +1,122 @@
+import net.fabricmc.fabric.impl.ingredient.builtin.NbtIngredient;
+import net.fabricmc.fabric.impl.ingredient.builtin.OrIngredient;
+
+@ApiStatus.Internal
```
Remove.

apple502j commented on fabric-ingredient-api-v1/src/main/java/net/fabricmc/fabric/impl/ingredient/ShapelessMatch.java:
```diff
@@ -0,0 +1,96 @@
+ * <p>The problem to solve is a maximum cardinality bipartite matching, for which this implementation uses the augmenting path algorithm.
+ * This has good performance in simple cases, and sufficient O(N^3) asymptotic complexity in the worst case.
+ */
+@ApiStatus.Internal
```
Remove.

apple502j commented on fabric-ingredient-api-v1/src/main/java/net/fabricmc/fabric/impl/ingredient/builtin/AndIngredient.java:
```diff
@@ -0,0 +1,71 @@
+
+import net.fabricmc.fabric.api.ingredient.v1.CustomIngredientSerializer;
+
+@ApiStatus.Internal
```
Remove.

apple502j commented on fabric-ingredient-api-v1/src/main/java/net/fabricmc/fabric/impl/ingredient/builtin/AndIngredient.java:
```diff
@@ -0,0 +1,71 @@
+	public ItemStack[] getMatchingStacks() {
+		List<ItemStack> previewStacks = new ArrayList<>();
+
+		if (ingredients.length > 0) {
```
Seems unnecessary due to check below. Comment instead?

apple502j commented on fabric-ingredient-api-v1/src/main/java/net/fabricmc/fabric/impl/ingredient/builtin/CombinedIngredient.java:
```diff
@@ -0,0 +1,115 @@
+	}
+
+	@Override
+	public boolean requiresTesting() {
```
I believe this is incorrect, at least for OR.

If ingredient A does not require testing and ingredient B requires testing, the whole ingredient needs testing, otherwise B is never matched.

Technici4n replied:
🤦

apple502j commented on fabric-ingredient-api-v1/src/main/java/net/fabricmc/fabric/api/ingredient/v1/DefaultCustomIngredients.java:
```diff
@@ -0,0 +1,134 @@
+
+	/**
+	 * Create an ingredient that matches the passed template stack, NBT included.
+	 * This check can either be strict (the exact NBT must match) or non-strict (the template NBT must be a subset of the stack NBT).
```
Document that this crashes with NBT-less stack and non-strict mode. Or rework the specs on this.

apple502j commented on fabric-ingredient-api-v1/src/main/java/net/fabricmc/fabric/impl/ingredient/builtin/OrIngredient.java:
```diff
@@ -0,0 +1,66 @@
+
+import net.fabricmc.fabric.api.ingredient.v1.CustomIngredientSerializer;
+
+@ApiStatus.Internal
```
Remove.

apple502j commented on fabric-ingredient-api-v1/src/main/resources/fabric-ingredient-api-v1.accesswidener:
```diff
@@ -0,0 +1,5 @@
+accessWidener   v1  named
```
```suggestion
accessWidener   v2  named
```

apple502j commented on fabric-ingredient-api-v1/src/main/resources/fabric-ingredient-api-v1.mixins.json:
```diff
@@ -0,0 +1,12 @@
+{
+  "required": true,
+  "package": "net.fabricmc.fabric.mixin.ingredient",
+  "compatibilityLevel": "JAVA_16",
```
```suggestion
  "compatibilityLevel": "JAVA_17",
```

apple502j commented on fabric-ingredient-api-v1/src/main/resources/fabric.mod.json:
```diff
@@ -0,0 +1,32 @@
+  ],
+  "accessWidener": "fabric-ingredient-api-v1.accesswidener",
+  "depends": {
+    "fabricloader": ">=0.14.9"
```
```suggestion
    "fabricloader": ">=0.14.11"
```

Technici4n replied:
I just downgraded to `0.14.10` because that's what Fabric API is using at the moment.

apple502j commented on fabric-ingredient-api-v1/src/main/resources/fabric.mod.json:
```diff
@@ -0,0 +1,32 @@
+  "depends": {
+    "fabricloader": ">=0.14.9"
+  },
+  "description": "Creation of custom ingredients",
```
Should be more descriptive.

Technici4n committed:
Apply suggestions from code review

Co-authored-by: apple502j <33279053+apple502j@users.noreply.github.com>

Technici4n committed:
Address 🍎 review

Technici4n committed:
Get rid of the @Overwrite

Technici4n committed:
Implement graceful fallback for clients not supporting some custom ingredients

Technici4n commented:
I just implemented a graceful fallback mechanism to make this API work nicely for clients that don't support (some or all) custom ingredients. Summary follows:
- On login, server requests the client's available ingredients.
- The client answers by listing the serializer IDs it supports.
- The server stores that where it's relevant.
- When an ingredient is serialized, we lookup which IDs are supported by the client. If we can lookup and the client doesn't support the ingredient, we send it as a vanilla stack list ingredient instead. Otherwise, we send the custom ingredient as usual.

List of cases to test:
- [x] Server has ingredient API, client doesn't have the API.
- [x] Server has ingredient API, client has the API but doesn't support some ingredients.
- [x] Server has ingredient API, client has the API and all the used ingredients.

Technici4n commented:
After some consideration, this new module will be renamed Fabric Recipe API v1, and the ingredient stuff will be moved to the `net.fabricmc.fabric.api.recipe.v1.ingredient` package, in preparation for possible other kinds of recipe format extensions.

Technici4n committed:
Move custom ingredient code to new Recipe API module

Technici4n commented:
![image](https://user-images.githubusercontent.com/13494793/208303274-bbfd2db0-3a12-4fe1-803d-ad3536c107d2.png)

Works fine on vanilla clients now (ignoring the recipe book) !

Technici4n committed:
Fix client package change

Technici4n commented:
Partial support on clients also works: (debug output on server, removed some ingredients on client)
```
[15:28:36] [Server thread/INFO]: Technici4n joined the game
Wrote fabric:nbt as a vanilla ingredient
Wrote fabric:and as a custom ingredient
Wrote fabric:difference as a vanilla ingredient
Wrote fabric:nbt as a vanilla ingredient
Wrote fabric:or as a custom ingredient
```

modmuss50 commented on fabric-recipe-api-v1/src/main/java/net/fabricmc/fabric/api/recipe/v1/ingredient/DefaultCustomIngredients.java:
```diff
@@ -0,0 +1,143 @@
+	 *
+	 * @throws IllegalArgumentException if the array is empty
+	 */
+	public static Ingredient and(Ingredient... ingredients) {
```
Nit: Maybe call this `all`

Technici4n replied:
Could also be `intersection`, not sure what's best.

modmuss50 replied:
Java's Stream class uses `allMatch` and `anyMatch` I think it would make sense to follow it's lead. I dont think the `match` part is needed in this context however?

modmuss50 commented on fabric-recipe-api-v1/src/main/java/net/fabricmc/fabric/api/recipe/v1/ingredient/DefaultCustomIngredients.java:
```diff
@@ -0,0 +1,143 @@
+	 *
+	 * @throws IllegalArgumentException if the array is empty
+	 */
+	public static Ingredient or(Ingredient... ingredients) {
```
Nit: Maybe call this `any`?

Technici4n replied:
Could also be `union`, not sure what's best.

62832 replied:
I'd argue for `any`/`all` since they're simpler. `union` and `intersection` seem a bit out of place as more set-theoretical terms.

modmuss50 commented on fabric-recipe-api-v1/src/main/java/net/fabricmc/fabric/impl/recipe/ingredient/CustomIngredientImpl.java:
```diff
@@ -0,0 +1,143 @@
+	public static void registerSerializer(CustomIngredientSerializer<?> serializer) {
+		Objects.requireNonNull(serializer.getIdentifier(), "CustomIngredientSerializer identifier may not be null.");
+
+		if (REGISTERED_SERIALIZERS.putIfAbsent(serializer.getIdentifier(), serializer) != null) {
+			throw new IllegalArgumentException("CustomIngredientSerializer with identifier " + serializer.getIdentifier() + " already registered.");
+		}
```
Nit: check if the key exists before adding, as a caller can catch the `IllegalArgumentException` to overwrite.

Technici4n replied:
That's why it's a `putIfAbsent`. :smile:

modmuss50 replied:
Ah yeah, good point. This is fine 👍

modmuss50 commented on fabric-recipe-api-v1/src/main/java/net/fabricmc/fabric/impl/recipe/ingredient/CustomIngredientImpl.java:
```diff
@@ -0,0 +1,143 @@
+	}
+
+	static {
+		CustomIngredientSerializer.register(AndIngredient.SERIALIZER);
```
Question, is this OK class loading wise? Might be better moved to an init entrypoint to be sure its registered?

Technici4n replied:
It works fine because the various ingredients don't reference `CustomIngredientImpl` (as they're using only the API).

modmuss50 replied:
Ah, this is fine. just seen that `CustomIngredientSerializer.get` actually calls `CustomIngredientImpl.getSerializer`. 👍 Might still be a tiny bit cleaner in the mod entrypoint, but not a major issue.

modmuss50 commented on fabric-recipe-api-v1/src/main/java/net/fabricmc/fabric/impl/recipe/ingredient/CustomIngredientImpl.java:
```diff
@@ -0,0 +1,143 @@
+		// We don't want to resolve the matching stacks,
+		// as this might cause the ingredient to use outdated tags when it's done too early.
+		// So we just return false when the matching stacks haven't been resolved yet (i.e. when the field is null).
+		// TODO: this is a bit hacky, can we not do better?
```
When is this called to early? Would hard crashing be better?

Technici4n replied:
![image](https://user-images.githubusercontent.com/13494793/208391470-89dc96c5-8fea-4ac0-a73c-61895ab4389d.png)

For example when used in a shapeless recipe. Vanilla tries to filter out empty ingredients for some reason, but tags haven't been bound yet, so the ingredient should not be resolved.

I don't know if we need to do anything about this, and in my testing this was never really an issue, so I'd suggest we leave it as-is for now. I removed the TODO.

modmuss50 commented on fabric-recipe-api-v1/src/main/java/net/fabricmc/fabric/impl/recipe/ingredient/CustomIngredientSync.java:
```diff
@@ -0,0 +1,87 @@
+ */
+public class CustomIngredientSync implements ModInitializer {
+	public static final Identifier PACKET_ID = new Identifier("fabric", "custom_ingredient_sync");
+	public static final int PROTOCOL_VERSION = 0;
```
Nit: Start with 1 for the first version? Not a major issue though.

Technici4n committed:
Address review comments

- And/Or -> All/Any
- Move builtin ingredient registration to entrypoint
- Initial protocol version is 1
- Misc other changes

Technici4n committed:
Add testing instructions

Technici4n committed:
Use a List for `getMatchingStacks`

Technici4n commented:
Just rebased to fix the `gradle.properties` conflict, and switched `getMatchingStacks` to a `List<ItemStack>`.

Technici4n commented on fabric-recipe-api-v1/src/main/java/net/fabricmc/fabric/impl/recipe/ingredient/CustomIngredientSync.java:
```diff
@@ -0,0 +1,87 @@
+ */
+public class CustomIngredientSync implements ModInitializer {
+	public static final Identifier PACKET_ID = new Identifier("fabric", "custom_ingredient_sync");
+	public static final int PROTOCOL_VERSION = 1;
```
We should rather have the server send the max supported protocol id, and the client answer back with the protocol it chooses. (e.g. in the future if we add protocol version 2 the server will send 1 or 2 depending on whether it supports it, and the client will reply with the highest protocol version that itself and the server support).

Technici4n replied:
Done

Technici4n committed:
Overengineer ingredient query a bit

modmuss50 commented:
I expect out of scope for this PR, but having a condition for a custom ingredient might be useful.

modmuss50 commented on fabric-recipe-api-v1/src/main/java/net/fabricmc/fabric/mixin/recipe/ingredient/PacketEncoderMixin.java:
```diff
@@ -0,0 +1,77 @@
+			method = "encode(Lio/netty/channel/ChannelHandlerContext;Lnet/minecraft/network/Packet;Lio/netty/buffer/ByteBuf;)V"
+	)
+	private void capturePacketEncoder(ChannelHandlerContext channelHandlerContext, Packet<?> packet, ByteBuf byteBuf, CallbackInfo ci) {
+		CustomIngredientSync.CURRENT_SUPPORTED_INGREDIENTS.set(fabric_supportedCustomIngredients);
```
Does this need to be done for all packets? Checking the Packet is an instance of the specific recipe packet may be better?

Technici4n replied:
The instanceof check will likely be more expensive than the ThreadLocal set.

modmuss50 reviewed (approved):
Look great. Before merging I will make sure to thoroughly test the networking stuff with vanilla and modded clients.

modmuss50 committed:
Merge remote-tracking branch 'origin/1.19.3' into custom-ingredient-api

# Conflicts:
#	gradle.properties
